src/transformers/trainer.py

```diff
@@ -435,6 +436,16 @@ def __init__(
                 )
             self.model_init = model_init
 
+        if is_deepspeed_zero3_enabled():
+            # Will reach this branch if the user has
+            # 1. Used `.from_pretrained` or `.from_config` to initialize their model
+            # 2. Did not configure Zero-3 via `TrainingArguments` or `accelerate launch` beforehand
+            # New models init such as `MyModel()` will not hit this step
+            if hasattr(model, "transformers_zero3_init_used") and not model.transformers_zero3_init_used:
+                raise ValueError(
+                    "Model was not initialized with `Zero-3` despite being configured. Please re-initialize your model via `***Model.from_pretrained(...)` or `***Model.from_config(...)` after creating your `TrainingArguments`!"
```

Comment:
I'd just get rid of the stars here as we could also have `***Model***.from_pretrained` etc., which I think can start to be confusing (just an opinion though and happy to be ignored here)

```suggestion
                    "Model was not initialized with `Zero-3` despite being configured. Please re-initialize your model via `from_pretrained(...)` or `from_config(...)` after creating your `TrainingArguments`!"
```